[UnifiedPDF][iOS] Relative scroll position is not preserved when resizing web view

pxlcoder opened the pull request:
#### 4312171edbbfbbfc2aa88a125eea2aeef57acbe2
<pre>
[UnifiedPDF][iOS] Relative scroll position is not preserved when resizing web view
<a href="https://bugs.webkit.org/show_bug.cgi?id=293906">https://bugs.webkit.org/show_bug.cgi?id=293906</a>
<a href="https://rdar.apple.com/152160616">rdar://152160616</a>

Reviewed by Richard Robinson, Wenson Hsieh, and Tim Horton.

When the web view is resized, the fitting scale of the PDF changes, resulting
in an individual page becoming taller or shorter, and an overall change to the
content size. However, the scroll view&apos;s content offset remains unchanged,
resulting in a &quot;jump&quot; to another page when the web view is resized.

Fix by computing a new content offset when the fitting scale changes. Since the
fitting scale is not actually exposed to the UI process, simply take the ratio
of new content height to the old content height.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):

Repurpose logic that pins the scroll position for traditional web content to
also support pinning the PDF scroll position.

Note that the added logic is only correct for PDFs that display vertically.
This is fine since WebKit does not support horizontal PDF display.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST(KeepRelativeScrollPositionAfterAnimatedResize)):
(TestWebKitAPI::UNIFIED_PDF_TEST(ScrollOffsetResetWhenChangingPDF)):

Canonical link: <a href="https://commits.webkit.org/295715@main">https://commits.webkit.org/295715@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44f36cb6f32e7a4ddb77c46040661c305093080e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105908 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25659 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16053 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111106 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56506 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107949 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34162 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80450 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108914 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20712 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95578 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60770 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13676 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55944 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90100 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13712 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113957 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33048 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24378 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89528 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33412 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91807 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89206 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34069 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11881 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28585 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17179 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32973 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38384 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32719 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36068 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34317 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1020,8 +1021,16 @@ - (void)_updateScrollViewForTransaction:(const WebKit::RemoteLayerTreeTransactio
-        LOG_WITH_STREAM(VisibleRects, stream << " updating scroll view with pageScaleFactor " << layerTreeTransaction.pageScaleFactor());
+
+    BOOL shouldUpdateZoomScale = !layerTreeTransaction.scaleWasSetByUIProcess() && ![_scrollView isZooming] && ![_scrollView isZoomBouncing] && ![_scrollView _wk_isZoomAnimating] && !WebKit::scalesAreEssentiallyEqual([_scrollView zoomScale], layerTreeTransaction.pageScaleFactor());
+    BOOL shouldUpdateContentOffsetForFittingScale = !shouldUpdateZoomScale && !CGSizeEqualToSize(oldContentSize, newContentSize) && (oldContentSize.height > 0) && self._isDisplayingPDF;
```
should the `oldContentSize.height > 0` just be `oldContentSize.height` because of the style rule not to compare to 0?

whsieh replied:
The "compare to 0" style rule applies to [equality checks](https://webkit.org/code-style-guidelines/#zero-comparison)

pxlcoder replied:
Technically it would be fine since content size shouldn't be negative, but I also don't see a strong reason to change it, so keeping as-is.

pxlcoder committed:
[UnifiedPDF][iOS] Relative scroll position is not preserved when resizing web view
https://bugs.webkit.org/show_bug.cgi?id=293906
rdar://152160616

Reviewed by Richard Robinson, Wenson Hsieh, and Tim Horton.

When the web view is resized, the fitting scale of the PDF changes, resulting
in an individual page becoming taller or shorter, and an overall change to the
content size. However, the scroll view's content offset remains unchanged,
resulting in a "jump" to another page when the web view is resized.

Fix by computing a new content offset when the fitting scale changes. Since the
fitting scale is not actually exposed to the UI process, simply take the ratio
of new content height to the old content height.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):

Repurpose logic that pins the scroll position for traditional web content to
also support pinning the PDF scroll position.

Note that the added logic is only correct for PDFs that display vertically.
This is fine since WebKit does not support horizontal PDF display.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST(KeepRelativeScrollPositionAfterAnimatedResize)):
(TestWebKitAPI::UNIFIED_PDF_TEST(ScrollOffsetResetWhenChangingPDF)):

Canonical link: https://commits.webkit.org/295715@main
